### PR TITLE
New: Buffer list

### PR DESCRIPTION
**Pull request by yo1995**
## Description

Implementation of the `Buffer list` sample. A brief demo below. 

Link to the branch folder: https://github.com/Esri/arcgis-runtime-samples-ios/tree/Ting/New-BufferList-vNext/arcgis-ios-sdk-samples/Geometry/Buffer%20list

|Demo (design is updated now)|
|------ |
|buffer-list-demo|

Please refer to `/runtime/common-samples/issues/539` for details.

### CODE REVIEW

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:160]: This loadable call can be moved into makeMap function.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (vquach2404) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:173]:
```suggestion
        // The spatial reference in this sample uses US feet as a unit.
```

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (vquach2404) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:180]:
```suggestion
        // Keep track of tapped points and their radii.
```

File: arcgis-ios-sdk-samples/Geometry/Buffer list/README.md
Comment (vquach2404) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/README.md:13]:
```suggestion
Tap on the map to add points. Tap the create button to draw buffer(s) around the points (the radius of the buffer is determined by the value of the slider). Toggle on union if you want the result to union (combine) the buffers. Tap the trash icon to start over. The red dashed envelope shows the area where you can expect reasonable results for planar buffer operations with the North Central Texas State Plane spatial reference.
```
Just a word suggestion.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/README.md
Comment (vquach2404) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/README.md:17]:
```suggestion
1. Use `class AGSGeometryEngine.bufferGeometries(_:distances:unionResults:)` to create a list of `AGSGeometry`s.
```
Consider making this plural.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (yo1995) [reply arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift]: Thanks! This is helpful. I don't know the plural form of radius before... 😅

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:82]:
```suggestion
    /// An array of tapped points and buffer radii (in US feet) tuple.
```

**Comment by vquach2404** [comment]
bufferList
After testing the sample some more, I realized that the radius doesn't update until after a new buffer has been created. Could you adjust it so that the value on the slider reflects the radius of the current buffer?

**Comment by yo1995** [comment]
> After testing the sample some more, I realized that the radius doesn't update until after a new buffer has been created. Could you adjust it so that the value on the slider reflects the radius of the current buffer?

Yeah I was unsure about the UI design. 🤔  The existing samples from dotnet and Java are both desktop, and they use a text field for the radius. I found it intrusive to present an alert asking for radius value, so I used slider instead. 

But a slider cannot explicitly tell which buffer the value is associated with. Do you mean to read the slider value after the center point graphic is created, and before adding the next point?

The better design I could come up with is to add a "preview buffer" for each center point, to visualize the size of the buffer. But that would make implementation details diverge.

Maybe I'll present this sample this Friday and ask for feedback.

**Comment by vquach2404** [comment]
> But a slider cannot explicitly tell which buffer the value is associated with. Do you mean to read the slider value after the center point graphic is created, and before adding the next point?

I just mean that the size of the current buffer that is being created should take the size that is reflected on the slider. So each time the value of the slider is changed, the radius should update to the new value.

**Comment by yo1995** [comment]
> I just mean that the size of the current buffer that is being created should take the size that is reflected on the slider. So each time the value of the slider is changed, the radius should update to the new value.

I see. The original design doc is a bit misleading. The create button is actually "drawing" the graphics, while "creating" the buffers is done by tapping on the map. The sample is designed to be used as repeating "set a slider value -> tap on the map" multiple times, and tap "create" once, to create a list of buffers.

I understand you are trying to use the create button as an add button, so that each time it's tapped, it add a buffer with the slider radius and the center point. I also think this workflow make more sense, but we'll need to consult our team if we want to change the design.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:133]: Clear button and create button should have some UI binding

- After clear is tapped, set both to disabled
- When more than 0 buffer exists, enable both buttons

**Comment by yo1995** [comment]
I'm moving this PR back to draft, and will discuss with our team on Friday about some design issues mentioned above.

**Comment by yo1995** [comment]
I've addressed some comments from Friday's review and decide to use the textField UI. However, since the deisgn is still being refactored by Trevor, I'll put it on hold for a while.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:62]: Since this overlay depends on the geometry of the `boundaryPolygon`, it is better to be instantiated by an instance method. However, I found it looking more "symmetric" to also create it as a computed variable. Wait for Phil's thought.

**Review by yo1995 — COMMENTED** [review]
@vquach2404 Please take another look as we are moving forward with the current datasource! Thanks.

**Review by vquach2404 — COMMENTED** [review]
simulator_screenshot_3A91E2B2-679A-46FA-9E0E-28F0194D1D6E
Please add some padding in between the Union label and the switch.

**Comment by vquach2404** [comment]
What do you think about creating the buffers immediately after the user enters its size? The Union switch could also recreate the buffers if the user changes the switch after the buffers have been made. That way the user can visualize how large the buffer will be and you might even be able to include an "undo last point" button.

**Comment by yo1995** [comment]
> What do you think about creating the buffers immediately after the user enters its size? The Union switch could also recreate the buffers if the user changes the switch after the buffers have been made. That way the user can visualize how large the buffer will be and you might even be able to include an "undo last point" button.

In this case each time the user tap on the map or toggle the switch, the whole buffer will get re-created, which might be not efficient. I'll make another commit later today to give it a try.

**Review by vquach2404 — COMMENTED** [review]
Thanks for making those changes! I think it looks great now and I feel like this workflow is much more intuitive. 

Also, is the cancel button supposed to be blue?
Simulator Screen Shot - iPhone 11 Pro Max - 2020-10-14 at 16 44 34

**Comment by yo1995** [comment]
> Also, is the cancel button supposed to be blue?

It should be purple. It seems only to be blue on iOS 14. 
Will need to consult Phil to see why it is blue when created... 🤔

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:79]: The plural of radius is _radii_ or _radiuses_.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:150]: Instead of iterating over `tappedPointsAndRadius` multiple times, I would suggest iterating over it once to get the array of points and radii:

```suggestion
        let (points, radii) = tappedPointsAndRadius.reduce(into: ([AGSPoint](), [NSNumber]())) { (result, pointAndRadius) in
            result.0.append(pointAndRadius.point)
            result.1.append(pointAndRadius.radius as NSNumber)
        }
        if let bufferPolygon = AGSGeometryEngine.bufferGeometries(points, distances: radii, unionResults: isUnionSwitch.isOn) {
            // Add graphics symbolizing the tap point.
            tappedLocationsGraphicsOverlay.graphics.addObjects(from: points.map { AGSGraphic(geometry: $0, symbol: nil) })
```

FYI: Double can be unconditionally cast to `NSNumber` (`double as NSNumber`, `doubles as [NSNumber]`, etc.). I don't know if casting is more efficient that initializing, but it is the preferred approach.

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:187]: I'm not sure it makes sense to split that sentence across the title and the message. It might be better to just combine them into a single sentence and use that as the title. Don't forget to add a period to the end!

image

File: arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift:89]: Correct me if I'm wrong, but it seems like this property isn't really necessary. The unit is used in one place, but the unit is always the same, so you could either declare a constant for it or just use `UnitLength.miles` directly. The value is only referenced in `doneAction`'s handler, so there really isn't a reason to store it in a property.

**Review by philium — APPROVED** [review]

**Comment by yo1995** [comment]
@saratchandrakarumuri Please take a look at this new sample 😁 .
@philium Sorry for dismissing your review but I had to solve the merge conflict. Please give another approval! Thanks.

**Review by philium — CHANGES_REQUESTED** [review]
The merged caused some items to become unordered, so Xcode automatically makes some changes to the project file when it is saved. Please make those changes in this PR.

FYI: Xcode keeps items sorted by identifier, so seeing items identifiers out-of-order is a clue that the project file needs to be re-saved.

image

**Comment by yo1995** [comment]
> The merged caused some items to become unordered...

Ah my bad. 🤦  I finally find out it is GitHub online merge solving editor that makes Xcode unhappy. Next time I'll proceed with more caution.

cc @vquach2404 in the future when resolving merge conflict in GitHub online editor, be sure to let Xcode perform the auto adjustments before merging the PR to `v.next`.

**Review by philium — APPROVED** [review]

**Review by saratchandrakarumuri — APPROVED** [review]
Looks good

**Pull request merged**
